### PR TITLE
Use the new function registry for the subscript functions.

### PR DESCRIPTION
**Pull request by kovrus**
## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java [hunk 1]
```diff
@@ -57,14 +57,73 @@
...
+        // subscript(array(any)), text) -> any
+        module.register(
+            Signature
+                .scalar(
+                    NAME,
+                    parseTypeSignature("array(E)"),
+                    DataTypes.INTEGER.getTypeSignature(),
+                    parseTypeSignature("E"))
+                .withTypeVariableConstraints(
+                typeVariable("E")),
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java:86]: nitpick: should be folded into the line above or indented.

File: server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java [hunk 2]
```diff
@@ -57,14 +57,73 @@
...
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNDEFINED.getTypeSignature()),
+            (signature, dataTypes) ->
+                new SubscriptFunction(
+                    signature,
+                    new FunctionInfo(new FunctionIdent(NAME, dataTypes), DataTypes.UNDEFINED),
+                    SubscriptFunction::lookupByName
+                )
+        );
+        module.register(
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java:108]: Could we document this a bit (why is this needed)?

File: server/src/test/java/io/crate/expression/scalar/SubscriptFunctionTest.java [hunk 3]
```diff
@@ -69,9 +69,7 @@ public void testIndexOutOfRange() throws Exception {
 
     @Test
     public void testIndexExpressionIsNotInteger() throws Exception {
-        expectedException.expectMessage(
-            "`index` in subscript expression (`base[index]`) " +
-            "must be a numeric type if the base expression is text_array");
+        expectedException.expectMessage("No cast function found for return type undefined");
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/SubscriptFunctionTest.java:72]: I think this error message is not really understandable for users. Is it possible to improve here?

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -401,7 +402,11 @@ private FrameBoundDefinition convertToAnalyzedFrameBound(ExpressionAnalysisConte
         int size = symbolsToCast.size();
         List<Symbol> castList = new ArrayList<>(size);
         for (int i = 0; i < size; i++) {
-            castList.add(symbolsToCast.get(i).cast(targetTypes.get(i), false));
+            var symbolToCast = symbolsToCast.get(i);
+            if (ArrayType.unnest(symbolToCast.valueType()).id() != ObjectType.ID) {
+                symbolToCast = symbolToCast.cast(targetTypes.get(i), false);
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:408]: Why are arrays of objects skipped here? Shouldn't the array-type/object-type be adapted instead?

File: server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java [hunk 5]
```diff
@@ -57,14 +57,73 @@
...
+                NAME,
+                parseTypeSignature("array(object)"),
+                DataTypes.STRING.getTypeSignature(),
+                parseTypeSignature("array(undefined)")),
+            (signature, dataTypes) -> new SubscriptFunction(
+                signature,
+                FunctionInfo.of(NAME, dataTypes, new ArrayType<>(DataTypes.UNDEFINED)),
+                SubscriptFunction::lookupIntoListObjectsByName)
+        );
+        // subscript(array(any)), text) -> any
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java:77]:
```suggestion
        // subscript(array(any)), integer) -> any
```

?

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 4 above)
Comment (kovrus) [reply server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: The description of this point from the chat communication 

```
The second issue is related to subscript on object or arrays of objects. There we try to infer the return type of the function at the normalization step based on arguments, see https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java#L98, but with the new function signature, the object arguments were cast to objects coz of different inner types (empty vs actual inner types) and the return type inference logic was not working any longer.
```

so basically I did that to skip cast for `object` and `array(object)` i am not sure how to handle in another way. hmm maybe actually checking whether the source and target are either `array(object)` and `object` and then don't cast source type if the source and target types have different inner types?

File: server/src/main/java/io/crate/expression/symbol/Symbol.java [hunk 6]
```diff
@@ -59,6 +61,9 @@ public final Symbol cast(DataType<?> targetType) {
     public Symbol cast(DataType<?> targetType, boolean tryCast) {
         if (targetType.equals(valueType())) {
             return this;
+        } else if (ArrayType.unnest(targetType).equals(DataTypes.UNTYPED_OBJECT)
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/symbol/Symbol.java:64]: so basically the cast for the `object` and `array(object)` was added coz the target type, the one that comes from the signature, has the empty inner types while the source data type has some inner types. There is no point to cast an object to an object type with empty inner types, so probably these semantic is not correct. I've added this condition in the `Symbol#cast`. @mfussenegger wdyt?

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/symbol/Symbol.java:64]: Looks better to me to have it here :+1:

File: server/src/test/java/io/crate/expression/scalar/SubscriptFunctionTest.java (hunk 3 above)
Comment (kovrus) [reply server/src/test/java/io/crate/expression/scalar/SubscriptFunctionTest.java]: probably cannot do it better than this https://github.com/crate/crate/pull/9928/files#diff-fb3a57a840597e5dc05aa82f387d13e6R406, otherwise, we will get the casting exception message.

Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/SubscriptFunctionTest.java]: 👍 Think this is much better now.

**Comment by kovrus** [comment]
@seut do you want to have a look at these changes?
